src/OpenIddict.EntityFrameworkCore/Stores/OpenIddictEntityFrameworkCoreAuthorizationStore.cs

```diff
@@ -661,38 +638,44 @@ public virtual IAsyncEnumerable<TAuthorization> ListAsync(int? count, int? offse
     /// <inheritdoc/>
     public virtual async ValueTask PruneAsync(DateTimeOffset threshold, CancellationToken cancellationToken)
     {
-        // Note: Entity Framework Core doesn't support set-based deletes, which prevents removing
-        // entities in a single command without having to retrieve and materialize them first.
+#if SUPPORTS_BULK_DBSET_OPERATIONS
+        // To prevent an SQL exception from being thrown if a new associated token is
+        // created after the existing tokens have been removed, the following logic is
+        // executed in a serializable transaction, that will lock the affected tables.
```

Comment:
Ah cool, you've got around to do it before me 😁 
I assume you mean that deletions of authorizations would fail if a new token entry to an authorization were added between deleting Tokens and Authorizations? Does `!authorization.Tokens.Any()` not sufficiently guard against that?

Bumping isolation all the way up to Serializable might guard against correctness issues such as trying to delete parents who've just acquired new childrens but at the cost of drastically increased chance of running into deadlocks and inducing massive blocking on those tables.

Especially since the two deletes have no Take on them, which means that once i would turn that task back on in my system (haven't turned the task back on since #1508), it would attempt to delete millions of rows in a single serializable transaction which... is bad xD I suppose under normal operation the task runs frequently enough to not delete too many rows at once but it's something to think about.

Reply:
Could we consider a .Take(1000).ExecuteDeleteAsync() ?

Reply:
> Bumping isolation all the way up to Serializable might guard against correctness issues such as trying to delete parents who've just acquired new childrens but at the cost of drastically increased chance of running into deadlocks and inducing massive blocking on those tables.

I agree it's not a perfect solution so I explored other options. One of them - implemented in the latest commit - uses `DeleteBehavior.SetNull` to let the database automatically set FKs to `null` without having to remove children entities on our side. Of course, this leaves orphaned authorizations and tokens when deleting an application or orphaned tokens when deleting an authorization, but I don't think it's a problem in this case (tho' it's certainly an observable behavior change).

Now, I'm not sure if `DeleteBehavior.SetNull` is compatible with all DBs: unlike `DeleteBehavior.Cascade` that doesn't work in this case with SQL Server (as there are cyclic references), `DeleteBehavior.SetNull` seems to work with at least both SQL Server and SQLite.

> Could we consider a .Take(1000).ExecuteDeleteAsync() ?

Certainly. I opted for 50K in my latest commit.

Let me know what you think 😄

Reply:
> One of them - implemented in the latest commit - uses `DeleteBehavior.SetNull` to let the database automatically set FKs to `null` without having to remove children entities on our side, which simplifies things a lot.

Another option I considered was to remove the constraints on foreign keys (after all, OpenIddict itself doesn't care much if FK columns point to inexistant/invalid entries), but it's not something EF Core seems to allow.

Reply:
@Suchiman @matthieume don't hesitate to share your thoughts 😃

Reply:
> unlike DeleteBehavior.Cascade that doesn't work in this case with SQL Server (as there are cyclic references)

If you set all FKs to Cascade, you'll run into multiple cascade path's but there are no cycles as far as i can see.
There are two cascade paths:
`Application -> Authorization -> Token`
`Application -> Token`
The thing that SQL Server is unhappy about is that deleting `Application` would try to delete `Token` twice in different paths. However we can see that the second path is unnecessary since the first path would take care of it.
So if you configure `Authorization.Application` and `Token.Authorization` to be `Cascade` and `Token.Application` to `NoAction`, it should work.

> Another option I considered was to remove the constraints on foreign keys (after all, OpenIddict itself doesn't care much if FK columns point to inexistant/invalid entries), but it's not something EF Core seems to allow.

It is possible, if you delete the ForeignKey constraint from the generated migration manually, EFC does not seem to complain about the FK values that do not exist in the database. However that is a very hard to sell user experience.

Reply:
> If you set all FKs to Cascade, you'll run into multiple cascade path's but there are no cycles as far as i can see.
> There are two cascade paths:
> `Application -> Authorization -> Token`
> `Application -> Token`
> The thing that SQL Server is unhappy about is that deleting `Application` would try to delete `Token` twice in different paths. However we can see that the second path is unnecessary since the first path would take care of it.
> So if you configure `Authorization.Application` and `Token.Authorization` to be `Cascade` and `Token.Application` to `NoAction`, it should work.

Interestingly, it's an approach I considered but it doesn't work when no intermediate `Authorization` entity is created... which is typically the case when the developer doesn't explicitly create a permanent authorization using the `IOpenIddictAuthorizationManager.CreateAsync()` API and prevents OpenIddict from creating ad-hoc authorizations using `options.DisableAuthorizationStorage()`. If you use `options.DisableAuthorizationStorage()`, the first cascade path doesn't kick in and you get an error since the tokens are not removed/detached:

```
SqlException : The DELETE statement conflicted with the REFERENCE constraint "FK_OpenIddictTokens_OpenIddictApplications_ApplicationId". The conflict occurred in database "openiddict-sandbox-aspnetcore-server", table "dbo.OpenIddictTokens", column 'ApplicationId'.
```

If we wanted to go this route, we'd have to remove `options.DisableAuthorizationStorage()` and always force the create of an authorization in the DB. It's definitely going to be a major breaking change and it removes an important scenario, so not sure it's our best option, sadly.

> It is possible, if you delete the ForeignKey constraint from the generated migration manually, EFC does not seem to complain about the FK values that do not exist in the database. However that is a very hard to sell user experience.

Ah yeah, when I meant it wasn't allowed, I meant via an entity configuration: as you say, forcing users to manually remove constraints from the migrations would be a terrible experience 😄

Reply:
I'm not at a computer right now so I can't test but is it possible to set `Token.Application` to SetNull while keeping the rest Cascade? Thinking about it though, do we have the same problems of people trying to delete applications while new tokens are being added? Can we workaround it by setting the application to disabled or trashing its client ID to prevent new tokens being created while running a retry loop that does { delete all application tokens; delete application}

Reply:
> I opted for 50K in my latest commit.

Can this be made configurable?  50k will lock the table and that can't happen in our deployment.

Reply:
I like the idea: I reverted the cascade changes I previously tested for now, but If I get reports indicating conflicts are a problem, I may end up implement the retry logic you suggest 😃